src/Compilers/Core/Portable/SourceGeneration/ISyntaxReceiver.cs

```diff
@@ -37,4 +40,41 @@ public interface ISyntaxReceiver
     /// </summary>
     /// <returns>An instance of an <see cref="ISyntaxReceiver"/></returns>
     public delegate ISyntaxReceiver SyntaxReceiverCreator();
+
+    /// <summary>
+    /// Receives notifications of each <see cref="SyntaxNode"/> in the compilation, along with a  
+    /// <see cref="SemanticModel"/> that can be queried to obtain more information, before generation
+    /// runs.
+    /// </summary>
+    /// <remarks>
+    /// A <see cref="ISourceGenerator"/> can provide an instance of <see cref="ISyntaxContextReceiver"/>
+    /// via a <see cref="SyntaxReceiverCreator"/>.
+    /// 
+    /// The compiler will invoke the <see cref="SyntaxReceiverCreator"/> prior to generation to 
+    /// obtain an instance of <see cref="ISyntaxContextReceiver"/>. This instance will have its 
+    /// <see cref="OnVisitSyntaxNode(GeneratorSyntaxContext)"/> called for every syntax node
+    /// in the compilation.
+    /// 
+    /// The <see cref="ISyntaxContextReceiver"/> can record any information about the nodes visited. 
+    /// During <see cref="ISourceGenerator.Execute(GeneratorExecutionContext)"/> the generator can obtain the 
+    /// created instance via the <see cref="GeneratorExecutionContext.SyntaxContextReceiver"/> property. The
+    /// information contained can be used to perform final generation.
+    /// 
+    /// A new instance of <see cref="ISyntaxContextReceiver"/> is created per-generation, meaning the instance
+    /// is free to store state without worrying about lifetime or reuse. 
+    /// 
+    /// An <see cref="ISourceGenerator"/> may provide only a single <see cref="ISyntaxReceiver"/> or
+    /// <see cref="ISyntaxContextReceiver" />, not both.
+    /// </remarks>
+    public interface ISyntaxContextReceiver
+    {
+        void OnVisitSyntaxNode(GeneratorSyntaxContext context);
+    }
+
+    /// <summary>
+    /// Allows a generator to provide instances of an <see cref="ISyntaxContextReceiver"/>
+    /// </summary>
+    /// <returns>An instance of an <see cref="ISyntaxContextReceiver"/></returns>
+    public delegate ISyntaxContextReceiver SyntaxContextReceiverCreator();
```

Comment:
Should the delegate return type be `ISyntaxContextReceiver?` instead since we handle `null` return values?

Reply:
Yeah, this was one of the questions I had about the null handling. We can explicitly allow `null`, ignore `null` values, or throw when provided with a `null`. The disadvantage of throwing is it won't occur until much later in the process, so seems disconnected. I think you're probably right that we should just explicitly allow `null` values via annotation. 

---
In reply to: [564862256](https://github.com/dotnet/roslyn/pull/50419#discussion_r564862256) [](ancestors = 564862256)